File: CSharp/StrategyCSharp/WidgetTest.cs
using System.Drawing;
using Moq;
using NUnit.Framework;

namespace StrategyCSharp
{
    public class WidgetTest
    {
        [Test]
        public void AWidgetCanBeCreated()
        {
            Widget w = new Widget();
        }

        [Test]
        public void AWidgetCanDescribeItself()
        {
            Widget w = new Widget();
            w.Describe();
        }

        [Test]
        public void AWidgetStartsAsATriangle()
        {
            Widget w = new Widget();
            Assert.True(w.Describe() == "Triangle");
        }

        [Test]
        public void AWidgetSTypeCanBeChangedAtRunTime()
        {
            Widget w = new Widget();
            Assert.True(w.Describe() == "Triangle");
            w.SetType(Type.Square);

            Assert.True(w.Describe() == "Square");
        }

        [Test]
        public void AWidgetCanSetToBeATriangle()
        {
            Widget w = new Widget();
            w.SetType(Type.Triangle);
        }

        [Test]
        public void ATriangularWidgetDescribesItselfAsATriangle()
        {
            Widget w = new Widget();
            w.SetType(Type.Triangle);

            Assert.True(w.Describe() == "Triangle");
        }

        [Test]
        public void ATriangularWidgetDrawsAsATriangle()
        {
            Widget w = new Widget();
            w.SetType(Type.Triangle);

            var c = new Mock<ICanvas>();

            w.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(It.IsAny<Point>()), Times.Exactly(3));
        }

        [Test]
        public void AWidgetCanSetToBeASquare()
        {
            Widget w = new Widget();
            w.SetType(Type.Square);
        }

        [Test]
        public void ASquareWidgetDescribesItselfAsASquare()
        {
            Widget w = new Widget();
            w.SetType(Type.Square);

            Assert.True(w.Describe() == "Square");
        }

        [Test]
        public void ASquareWidgetDrawsAsASquare()
        {
            Widget w = new Widget();
            w.SetType(Type.Square);

            var c = new Mock<ICanvas>();

            w.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(It.IsAny<Point>()), Times.Exactly(4));
        }

        [Test]
        public void AWidgetCanSetToBeAPentagon()
        {
            Widget w = new Widget();
            w.SetType(Type.Pentagon);
        }

        [Test]
        public void APentagonWidgetDescribesItselfAsAPentagon()
        {
            Widget w = new Widget();
            w.SetType(Type.Pentagon);

            Assert.True(w.Describe() == "Pentagon");
        }

        [Test]
        public void APentagonWidgetDrawsAsAPentagon()
        {
            Widget w = new Widget();
            w.SetType(Type.Pentagon);

            var c = new Mock<ICanvas>();

            w.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(It.IsAny<Point>()), Times.Exactly(5));
        }

        [Test]
        public void AWidgetCanSetToBeAHexagon()
        {
            Widget w = new Widget();
            w.SetType(Type.Hexagon);
        }

        [Test]
        public void AHexagonWidgetDescribesItselfAsAHexagon()
        {
            Widget w = new Widget();
            w.SetType(Type.Hexagon);

            Assert.True(w.Describe() == "Hexagon");
        }

        [Test]
        public void AHexagonWidgetDrawsAsAHexagon()
        {
            Widget w = new Widget();
            w.SetType(Type.Hexagon);

            var c = new Mock<ICanvas>();

            w.Draw(c.Object);

            c.Verify(mock => mock.DrawVertex(It.IsAny<Point>()), Times.Exactly(6));
        }
    }
}

File: Java/StrategyPattern/src/test/java/org/example/WidgetTest.java
package org.example;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

public class WidgetTest
{
    @Test
    public void aWidgetCanBeCreated()
    {
        var w = new Widget();
    }

    @Test
    public void aWidgetCanDescribeItself()
    {
        var w = new Widget();
        w.describe();
    }

    @Test
    public void aWidgetStartsAsATriangle()
    {
        var w = new Widget();
        assertEquals("Triangle", w.describe());
    }

    @Test
    public void aWidgetSTypeCanBeChangedAtRunTime()
    {
        var w = new Widget();
        assertEquals("Triangle", w.describe());
        w.setType(Type.Square);

        assertEquals("Square", w.describe());
    }

    @Test
    public void aWidgetCanSetToBeATriangle()
    {
        var w = new Widget();
        w.setType(Type.Triangle);
    }

    @Test
    public void aTriangularWidgetDescribesItselfAsATriangle()
    {
        var w = new Widget();
        w.setType(Type.Triangle);

        assertEquals("Triangle", w.describe());
    }

    @Test
    public void aTriangularWidgetDrawsAsATriangle()
    {
        VerifyCorrectNumberOfVertices(Type.Triangle, 3);
    }

    @Test
    public void aWidgetCanSetToBeASquare()
    {
        var w = new Widget();
        w.setType(Type.Square);
    }

    @Test
    public void aSquareWidgetDescribesItselfAsASquare()
    {
        var w = new Widget();
        w.setType(Type.Square);

        assertEquals("Square", w.describe());
    }

    @Test
    public void aSquareWidgetDrawsAsASquare()
    {
        VerifyCorrectNumberOfVertices(Type.Square, 4);
    }

    @Test
    public void aWidgetCanSetToBeAPentagon()
    {
        var w = new Widget();
        w.setType(Type.Pentagon);
    }

    @Test
    public void aPentagonWidgetDescribesItselfAsAPentagon()
    {
        var w = new Widget();
        w.setType(Type.Pentagon);

        assertEquals("Pentagon", w.describe());
    }

    @Test
    public void aPentagonWidgetDrawsAsAPentagon()
    {
        VerifyCorrectNumberOfVertices(Type.Pentagon, 5);
    }

    @Test
    public void aWidgetCanSetToBeAHexagon()
    {
        var w = new Widget();
        w.setType(Type.Hexagon);
    }

    @Test
    public void aHexagonWidgetDescribesItselfAsAHexagon()
    {
        var w = new Widget();
        w.setType(Type.Hexagon);

        assertEquals("Hexagon", w.describe());
    }

    @Test
    public void aHexagonWidgetDrawsAsAHexagon()
    {
        VerifyCorrectNumberOfVertices(Type.Hexagon, 6);
    }

    private static void VerifyCorrectNumberOfVertices(Type type, int vertices)
    {
        var w = new Widget();
        w.setType(type);

        var c = mock(ICanvas.class);

        w.draw(c);

        verify(c, times(vertices)).drawVertex(any(Point.class));
    }
}